### PR TITLE
Refactored Feat 2.2.4 basic table data formatting

### PR DESCRIPTION
**Pull request by Murray2015**

### CODE REVIEW

**Comment by maxmalynowsky** [comment]
I see there are some merge conflicts to be resolved from Feat 2.2.3 being merged into main.

**Comment by Murray2015** [comment]
@maxmalynowsky conflicts fixed

**Review by maxmalynowsky — CHANGES_REQUESTED** [review]

**Comment by maxmalynowsky** [comment]
I just noticed that the pre-commit isn't working properly, I had to change it to the following to get all the checks to work:

```yaml
fail_fast: true
repos:
  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.6.2
    hooks:
      - id: ruff
        args: [--fix]
      - id: ruff-format
  - repo: https://github.com/jackdewinter/pymarkdown
    rev: v0.9.22
    hooks:
      - id: pymarkdown
```

I also noticed I was putting the import sorting ruff rule in `.pre-commit-config.yaml` before, the full list in `pyproject.toml` should be:

```toml
[tool.ruff.lint]
select = ["ANN0", "D", "E4", "E7", "E9", "F", "I", "TID"]
```

File: tests/table_data_formatting/test_table_data_formatting_has_data.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from src.checks.table_data_formatting_has_data import main
+
+ISO = "mdg"
+
+
+def test_table_data_formatting_has_data(gdfs):
```
[the remark this answers is not in the document]
Comment (Murray2015) [reply tests/table_data_formatting/test_table_data_formatting_has_data.py:6]: Good spot, agreed!

File: src/utils.py [hunk 2]
```diff
@@ -13,6 +13,8 @@
 # Could do more with this type, as iso3 and levels keys are required.
 type CheckReturnList = list[dict[str, Any]]
 
+ADMIN_BOUNDARY_REGEX = r"^[aA][dD](?:M|I|N)*(\d).*"
```
[the remark this answers is not in the document]
Comment (Murray2015) [reply src/utils.py:16]: Hmm I see where you're going as they're not really "utils", but types and regex aren't really "config" either. What about making a `constants.py` to hold them?

[the remark this answers is not in the document]
Comment (maxmalynowsky) [reply src/utils.py:16]: Maybe, if we go about creating something like a `consts.py` file, I'm not sure what else currently in `config.py` should be added too. `TIMEOUT` has the option of being defined by an environment variable, and path locations like `tables` or column lists like `metadata_columns` could go in either one.

[the remark this answers is not in the document]
Comment (Murray2015) [reply src/utils.py]: Yeah you're right, might be overkill to add extra modules with just a couple of constants in them. I've added to config, we can factor constants out in to a separate module if it `config.py` starts growing unreasonably

File: src/checks/table_data_formatting_heirarchical.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    Returns:
+        List of check rows to be outputed as a CSV.
+    """
+    check_results = []
+    for admin_level, gdf in enumerate(gdfs):
+        row = {
+            "iso3": iso3,
+            "level": admin_level,
+            "proportion_of_expected_heirarchical_columns": proportion_of_expected_heirarchical_columns(
```
Comment (maxmalynowsky) [inline src/checks/table_data_formatting_heirarchical.py:24]: Do we want to leave the creation of a 0-1 score for the `scores` module? This check could be simplified if it only counts the total number of hierarchical columns:

```py
"total_heirarchical_columns": len(
    [
        any(column.startswith(f"ADM{level}") for column in gdf.columns)
        for level in range(admin_level + 1)
    ]
),
```

**Review by maxmalynowsky — CHANGES_REQUESTED** [review]

**Comment by maxmalynowsky** [comment]
Just comparing the check results of these two files, and is `table_data_formatting_has_data.py` needed? It looks like that check is the same as `proportion_of_expected_heirarchical_columns` == 0.

**Comment by maxmalynowsky** [comment]
I was taking a look back at the definition of 2.2.4 / 2.3.4, as well as reflecting on the columns we'd expect for an ADM0 in English (`ADM0_EN`,`ADM0_PCODE`, `ADM0_REF`, `ADM0ALT1_EN`, `ADM0ALT2_EN`). I think it would make sense to check for the existence of all those columns in this module. I think this would cover everything?

```py
row = {
    "iso3": iso3,
    "level": admin_level,
    "levels_with_data": sum(
        [
            any(column.startswith(f"ADM{level}") for column in gdf.columns)
            for level in range(admin_level + 1)
        ]
    ),
    "levels_with_name": sum(
        [
            any(
                bool(match(rf"^ADM{level}_[A-Z][A-Z]$", column))
                for column in gdf.columns
            )
            for level in range(admin_level + 1)
        ]
    ),
    "levels_with_pcode": sum(
        [
            any(column == f"ADM{level}_PCODE" for column in gdf.columns)
            for level in range(admin_level + 1)
        ]
    ),
    "levels_with_reference_name": sum(
        [
            any(column == f"ADM{level}_REF" for column in gdf.columns)
            for level in range(admin_level + 1)
        ]
    ),
    "levels_with_alternative_name": sum(
        [
            any(
                bool(match(rf"^ADM{level}ALT[1-2]_[A-Z][A-Z]$", column))
                for column in gdf.columns
            )
            for level in range(admin_level + 1)
        ]
    ),
    "name_count": sum(
        [
            bool(match(rf"^ADM{level}_[A-Z][A-Z]$", column))
            for column in gdf.columns
            for level in range(admin_level + 1)
        ]
    ),
    "alternative_name_count": sum(
        [
            bool(match(rf"^ADM{level}ALT[1-2]_[A-Z][A-Z]$", column))
            for column in gdf.columns
            for level in range(admin_level + 1)
        ]
    ),
}
```

File: src/config.py [hunk 4]
```diff
@@ -56,3 +57,10 @@
     "hdx_source_2",
     "hdx_license",
 ]
+
+
+# A return type for the checks in /checks.
+# Could do more with this type, as iso3 and levels keys are required.
+type CheckReturnList = list[dict[str, Any]]
+
+ADMIN_BOUNDARY_REGEX = r"^[aA][dD](?:M|I|N)*(\d).*"
```
[the remark this answers is not in the document]
Comment (Murray2015) [reply src/config.py:66]: Good question; I'm not sure. @leonbaruah / @ddebarros-mapaction, what's the intention here?

File: src/checks/table_data_formatting_heirarchical.py (hunk 3 above)
Comment (Murray2015) [reply src/checks/table_data_formatting_heirarchical.py:24]: You're right, we should prefer to leave the creation of scores to the `scores` module. I was lazy here, happy to refactor. That suggestion also simplifies the code too, thanks

**Comment by Murray2015** [comment]
> Just comparing the check results of these two files, is `table_data_formatting_has_data.py` needed? It looks like `has_at_least_1_data_column` is the same as `proportion_of_expected_heirarchical_columns` > 0.

I kept two separate functions, as the first is what is specified in the Confluence checks page, and the hierarchical version was me thinking that's a more useful check, assuming my understanding of the requirements for hierarchical columns is true. Do we know if the hierarchical assumption is correct?

**Comment by Murray2015** [comment]
> I was taking a look back at the definition of 2.2.4 / 2.3.4, as well as reflecting on the columns we'd expect for an ADM0 in English (`ADM0_EN`,`ADM0_PCODE`, `ADM0_REF`, `ADM0ALT1_EN`, `ADM0ALT2_EN`). I think it would make sense to check for the existence of all those columns in this module. I think this would cover everything?
> 
> ```python
> row = {
>     "iso3": iso3,
>     "level": admin_level,
>     "levels_with_data": sum(
>         [
>             any(column.startswith(f"ADM{level}") for column in gdf.columns)
>             for level in range(admin_level + 1)
>         ]
>     ),
>     "levels_with_name": sum(
>         [
>             any(
>                 bool(match(rf"^ADM{level}_[A-Z][A-Z]$", column))
>                 for column in gdf.columns
>             )
>             for level in range(admin_level + 1)
>         ]
>     ),
>     "levels_with_pcode": sum(
>         [
>             any(column == f"ADM{level}_PCODE" for column in gdf.columns)
>             for level in range(admin_level + 1)
>         ]
>     ),
>     "levels_with_reference_name": sum(
>         [
>             any(column == f"ADM{level}_REF" for column in gdf.columns)
>             for level in range(admin_level + 1)
>         ]
>     ),
>     "levels_with_alternative_name": sum(
>         [
>             any(
>                 bool(match(rf"^ADM{level}ALT[1-2]_[A-Z][A-Z]$", column))
>                 for column in gdf.columns
>             )
>             for level in range(admin_level + 1)
>         ]
>     ),
>     "name_count": sum(
>         [
>             bool(match(rf"^ADM{level}_[A-Z][A-Z]$", column))
>             for column in gdf.columns
>             for level in range(admin_level + 1)
>         ]
>     ),
>     "alternative_name_count": sum(
>         [
>             bool(match(rf"^ADM{level}ALT[1-2]_[A-Z][A-Z]$", column))
>             for column in gdf.columns
>             for level in range(admin_level + 1)
>         ]
>     ),
> }
> ```

I don't know the expected file formats, so I'm happy to defer to your expertise here! Cc. @leonbaruah / @ddebarros-mapaction, this is outside of the checks' definition but makes sense to me. Are you happy for us to proceed with this?

**Comment by maxmalynowsky** [comment]
> I don't know the expected file formats, so I'm happy to defer to your expertise here! Cc. @leonbaruah / @ddebarros-mapaction, this is outside of the checks' definition but makes sense to me. Are you happy for us to proceed with this?

I had a meeting with Leon yesterday and he mentioned that the confluence checks page is intended to be more of a starting point to guide development. We acknowledged that the document's checks are more of a mock-up and could be further refined for specificity. I didn't see any other check definition that covered the existence of all the expected columns, so I thought 2.2.4 would be a good place for it. Here's another example of the expected column format, using Afghanistan Admin 2. I think my suggestion above for how to score in 2.3.4 would cover everything.

- `ADM2_EN`: English name
- `ADM2_DA`: Dari name (`da` is actually the Danish language code, this is going to be changed to `fa` soon) 
- `ADM2_PCODE`: Admin 2 Place Code (P-Code)
- `ADM2_REF`: Takes the first language code (English in this case) and removes any special characters for database use. Should only be present for the current admin level (i.e. not present for ADM1 or ADM0).
- `ADM2ALT1_EN`: Alternative name place in English. Should only be present for the current admin level (i.e. not present for ADM1 or ADM0).
- `ADM2ALT2_EN`: Another alternative name place in English. Should only be present for the current admin level (i.e. not present for ADM1 or ADM0).
- `ADM2ALT1_DA`: Alternative place name in Dari. Should only be present for the current admin level (i.e. not present for ADM1 or ADM0).
- `ADM2ALT2_DA`: Another alternative place name in Dari. Should only be present for the current admin level (i.e. not present for ADM1 or ADM0).
- `ADM1_EN`: Admin 1 name in English.
- `ADM1_DA`: Admin 1 name in Dari.
- `ADM1_PCODE`: Admin 1 P-Code.
- `ADM0_EN`: Admin 0 name in English.
- `ADM0_DA`: Admin 0 name in Dari.
- `ADM0_PCODE`: Admin 0 P-Code.

**Comment by Murray2015** [comment]
@maxmalynowsky nice work on catching up with Leon - I've incorporated your suggestions above !

**Comment by maxmalynowsky** [comment]
@Murray2015, I see there's still a few merge conflicts with `main`.

**Comment by Murray2015** [comment]
@maxmalynowsky finally got around to this. Due to the long-lived branch, there were a lot of merge conflicts. Ugh.

**Comment by maxmalynowsky** [comment]
Yeah, and the pre-commit checks hadn't been working properly until now, so that must have added an extra level of changes. I'll take one last look now!

File: src/checks/table_data_formatting_heirarchical.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    Returns:
+        List of check rows to be outputed as a CSV.
+    """
+    check_results = []
+    for admin_level, gdf in enumerate(gdfs):
+        row = {
+            "iso3": iso3,
+            "level": admin_level,
+            "total_heirarchical_columns": sum(
```
Comment (maxmalynowsky) [inline src/checks/table_data_formatting_heirarchical.py:22]: I see this has the exact same check function as `levels_with_data` in `table_data_formatting_has_data.py`, so it should be safe to delete.

**Review by maxmalynowsky — CHANGES_REQUESTED** [review]

**Comment by maxmalynowsky** [comment]
It looks like there are just two small redundancies left.

File: src/config.py [hunk 6]
```diff
@@ -63,3 +63,10 @@
...
```
[the remark this answers is not in the document]
Comment (Murray2015) [reply src/config.py]: Fixed

File: src/checks/table_data_formatting_heirarchical.py (hunk 5 above)
Comment (Murray2015) [reply src/checks/table_data_formatting_heirarchical.py]: Also fixed

**Review by maxmalynowsky — APPROVED** [review]

**Pull request merged**